File: backend-go/projects/gateway/internal/openaicompat/computeradapter.go
package openaicompat

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// ComputerAdapter ports openai-compatible-computer/computer-adapter.ts: the
// configured HTTP browser adapter executor for OpenAI-hosted computer-use
// tools. Node gates it on hostedToolRuntimes.computer === 'local_runtime'
// plus computerAdapter.enabled + endpoint; the executor override replaces the
// Node set-openAICompatibleComputerExecutorForTest hook.

// ComputerCallResult mirrors OpenAIToAnthropicComputerCallResult.
type ComputerCallResult struct {
	CallID   string
	Status   string
	Actions  []map[string]any
	Metadata map[string]any
}

// ComputerRuntimeResult mirrors OpenAIToAnthropicComputerRuntimeResult.
type ComputerRuntimeResult struct {
	Message  string
	Call     *ComputerCallResult
	Metadata map[string]any
}

// ComputerRuntimeInput mirrors OpenAIToAnthropicComputerRuntimeInput.
type ComputerRuntimeInput struct {
	Body   map[string]any
	Tool   map[string]any
	Stream bool
}

// ComputerExecutor mirrors OpenAIToAnthropicComputerExecutor.
type ComputerExecutor interface {
	Run(ctx context.Context, input ComputerRuntimeInput) (*ComputerRuntimeResult, error)
}

type httpComputerExecutor struct {
	config ComputerAdapterConfig
	client HTTPDoer
}

// ComputerExecutorForRequest mirrors
// openAICompatibleComputerExecutorForGatewayRequest: nil unless the hosted
// tool runtime mode is local_runtime and the adapter is fully configured.
// override replaces the executor wholesale (the Node test hook).
func ComputerExecutorForRequest(config Config, client HTTPDoer, override ComputerExecutor) ComputerExecutor {
	if override != nil {
		return override
	}
	config = config.withDefaults()
	if config.HostedToolComputerMode != "local_runtime" {
		return nil
	}
	if !config.ComputerAdapter.Enabled || config.ComputerAdapter.Endpoint == "" {
		return nil
	}
	if client == nil {
		client = http.DefaultClient
	}
	return &httpComputerExecutor{config: config.ComputerAdapter, client: client}
}

// Run mirrors runConfiguredComputerHttpAdapter.
func (e *httpComputerExecutor) Run(ctx context.Context, input ComputerRuntimeInput) (*ComputerRuntimeResult, error) {
	endpoint := e.config.Endpoint
	if endpoint == "" {
		return nil, fmt.Errorf("Computer browser adapter endpoint is not configured")
	}
	timeoutCtx, cancel := context.WithTimeout(ctx, time.Duration(e.config.TimeoutMs)*time.Millisecond)
	defer cancel()
	body, err := json.Marshal(map[string]any{
		"body":   input.Body,
		"tool":   input.Tool,
		"stream": input.Stream,
	})
	if err != nil {
		return nil, computerTimeoutOrError(ctx, err)
	}
	request, err := http.NewRequestWithContext(timeoutCtx, http.MethodPost, endpoint, bytes.NewReader(body))
	if err != nil {
		return nil, computerTimeoutOrError(ctx, err)
	}
	request.Header.Set("accept", "application/json")
	request.Header.Set("content-type", "application/json")
	response, err := e.client.Do(request)
	if err != nil {
		if timeoutCtx.Err() != nil && ctx.Err() == nil {
			return nil, fmt.Errorf("Computer browser adapter request timed out")
		}
		return nil, err
	}
	defer func() {
		_, _ = io.Copy(io.Discard, io.LimitReader(response.Body, 1<<20))
		_ = response.Body.Close()
	}()
	text, err := readAdapterText(response, e.config.MaxBodyBytes)
	if err != nil {
		return nil, err
	}
	if response.StatusCode < 200 || response.StatusCode > 299 {
		return nil, fmt.Errorf("Computer browser adapter HTTP %d: %s", response.StatusCode, truncateString(text, 512))
	}
	var parsed any
	if jsonErr := json.Unmarshal([]byte(text), &parsed); jsonErr != nil {
		return nil, fmt.Errorf("Computer browser adapter response is not valid JSON")
	}
	return normalizeComputerAdapterResult(parsed)
}

// computerTimeoutOrError mirrors the abort timedOut branch.
func computerTimeoutOrError(ctx context.Context, err error) error {
	if ctx != nil && ctx.Err() != nil {
		return fmt.Errorf("Computer browser adapter request timed out")
	}
	return err
}

// readAdapterText mirrors readResponseTextWithLimit in the adapter module.
func readAdapterText(response *http.Response, maxBodyBytes int64) (string, error) {
	if response.Body == nil {
		return "", nil
	}
	buffer, err := io.ReadAll(io.LimitReader(response.Body, maxBodyBytes+1))
	if err != nil {
		return "", err
	}
	if int64(len(buffer)) > maxBodyBytes {
		return "", fmt.Errorf("Computer browser adapter response body exceeded limit")
	}
	return string(buffer), nil
}

func truncateString(text string, limit int) string {
	runes := []rune(text)
	if len(runes) <= limit {
		return text
	}
	return string(runes[:limit])
}

// normalizeComputerAdapterResult mirrors normalizeComputerAdapterResult.
func normalizeComputerAdapterResult(value any) (*ComputerRuntimeResult, error) {
	record, ok := value.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("Computer browser adapter response must be a JSON object")
	}
	message := "Computer browser adapter completed."
	if text := adapterString(record["message"]); text != nil {
		message = *text
	}
	result := &ComputerRuntimeResult{Message: message, Metadata: map[string]any{}}
	if callRecord := objectValue(record["call"]); callRecord != nil {
		call := &ComputerCallResult{Actions: []map[string]any{}}
		if callID := firstAdapterString(callRecord["call_id"], callRecord["callId"]); callID != nil {
			call.CallID = *callID
		}
		if status := adapterString(callRecord["status"]); status != nil {
			call.Status = *status
		}
		call.Actions = adapterActions(callRecord["actions"])
		if metadata := objectValue(callRecord["metadata"]); metadata != nil {
			call.Metadata = metadata
		}
		result.Call = call
	}
	if metadata := objectValue(record["metadata"]); metadata != nil {
		for key, value := range metadata {
			result.Metadata[key] = value
		}
	}
	result.Metadata["adapter"] = "http_browser"
	return result, nil
}

// adapterString mirrors stringValue in the adapter (trimmed non-empty).
func adapterString(value any) *string {
	text, ok := value.(string)
	if !ok {
		return nil
	}
	trimmed := strings.TrimSpace(text)
	if trimmed == "" {
		return nil
	}
	return &trimmed
}

func firstAdapterString(values ...any) *string {
	for _, value := range values {
		if text := adapterString(value); text != nil {
			return text
		}
	}
	return nil
}

// adapterActions mirrors arrayRecordValue (objects only).
func adapterActions(value any) []map[string]any {
	list, ok := value.([]any)
	if !ok {
		return []map[string]any{}
	}
	actions := []map[string]any{}
	for _, item := range list {
		if record, ok := item.(map[string]any); ok {
			actions = append(actions, record)
		}
	}
	return actions
}

File: migration-backup/node/final-archive/backend/src/modules/openai-compatible-computer/computer-adapter.ts
import type { Request } from 'express'

import { runtimeConfig } from '../../config/runtime.js'
import type {
  OpenAIToAnthropicComputerExecutor,
  OpenAIToAnthropicComputerRuntimeInput,
  OpenAIToAnthropicComputerRuntimeResult
} from '../providers/drivers/_shared/openai-anthropic-bridge.js'

let openAICompatibleComputerExecutorForTest: OpenAIToAnthropicComputerExecutor | undefined
const configuredComputerHttpExecutor: OpenAIToAnthropicComputerExecutor = {
  run: runConfiguredComputerHttpAdapter
}

export function setOpenAICompatibleComputerExecutorForTest(
  executor: OpenAIToAnthropicComputerExecutor | undefined
): void {
  openAICompatibleComputerExecutorForTest = executor
}

export function openAICompatibleComputerExecutorForGatewayRequest(
  _req: Request
): OpenAIToAnthropicComputerExecutor | undefined {
  if (openAICompatibleComputerExecutorForTest) return openAICompatibleComputerExecutorForTest
  if (runtimeConfig.hostedToolRuntimes.computer !== 'local_runtime') return undefined
  if (!runtimeConfig.computerAdapter.enabled || !runtimeConfig.computerAdapter.endpoint) return undefined
  return configuredComputerHttpExecutor
}

async function runConfiguredComputerHttpAdapter(
  input: OpenAIToAnthropicComputerRuntimeInput
): Promise<OpenAIToAnthropicComputerRuntimeResult> {
  const endpoint = runtimeConfig.computerAdapter.endpoint
  if (!endpoint) {
    throw new Error('Computer browser adapter endpoint is not configured')
  }
  const abort = createComputerAdapterAbortController(input.signal, runtimeConfig.computerAdapter.timeoutMs)
  try {
    const response = await fetch(endpoint, {
      method: 'POST',
      headers: {
        accept: 'application/json',
        'content-type': 'application/json'
      },
      body: JSON.stringify({
        body: input.body,
        tool: input.tool,
        stream: input.stream
      }),
      signal: abort.controller.signal
    })
    const text = await readResponseTextWithLimit(response, runtimeConfig.computerAdapter.maxBodyBytes)
    if (!response.ok) {
      throw new Error(`Computer browser adapter HTTP ${response.status}: ${text.slice(0, 512)}`)
    }
    let parsed: unknown
    try {
      parsed = JSON.parse(text)
    } catch {
      throw new Error('Computer browser adapter response is not valid JSON')
    }
    return normalizeComputerAdapterResult(parsed)
  } catch (error) {
    if (abort.timedOut()) {
      throw new Error('Computer browser adapter request timed out')
    }
    throw error
  } finally {
    abort.dispose()
  }
}

function normalizeComputerAdapterResult(value: unknown): OpenAIToAnthropicComputerRuntimeResult {
  if (!isRecord(value)) {
    throw new Error('Computer browser adapter response must be a JSON object')
  }
  const message = stringValue(value.message) ?? 'Computer browser adapter completed.'
  const callRecord = objectValue(value.call)
  return {
    message,
    call: callRecord ? {
      callId: stringValue(callRecord.call_id) ?? stringValue(callRecord.callId),
      status: stringValue(callRecord.status),
      actions: arrayRecordValue(callRecord.actions),
      metadata: objectValue(callRecord.metadata)
    } : undefined,
    metadata: {
      ...(objectValue(value.metadata) ?? {}),
      adapter: 'http_browser'
    }
  }
}

function createComputerAdapterAbortController(signal: AbortSignal | undefined, timeoutMs: number): {
  controller: AbortController
  timedOut: () => boolean
  dispose: () => void
} {
  const controller = new AbortController()
  let didTimeout = false
  const timeout = setTimeout(() => {
    didTimeout = true
    controller.abort()
  }, timeoutMs)
  const onAbort = () => controller.abort()
  signal?.addEventListener('abort', onAbort, { once: true })
  return {
    controller,
    timedOut: () => didTimeout,
    dispose: () => {
      clearTimeout(timeout)
      signal?.removeEventListener('abort', onAbort)
    }
  }
}

async function readResponseTextWithLimit(response: Response, maxBodyBytes: number): Promise<string> {
  if (!response.body) return await response.text()
  const reader = response.body.getReader()
  const chunks: Uint8Array[] = []
  let total = 0
  while (true) {
    const { done, value } = await reader.read()
    if (done) break
    if (!value) continue
    total += value.byteLength
    if (total > maxBodyBytes) {
      await reader.cancel()
      throw new Error('Computer browser adapter response body exceeded limit')
    }
    chunks.push(value)
  }
  return Buffer.concat(chunks.map((chunk) => Buffer.from(chunk))).toString('utf8')
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function objectValue(value: unknown): Record<string, unknown> | undefined {
  return isRecord(value) ? value : undefined
}

function stringValue(value: unknown): string | undefined {
  return typeof value === 'string' && value.trim() ? value.trim() : undefined
}

function arrayRecordValue(value: unknown): Array<Record<string, unknown>> {
  if (!Array.isArray(value)) return []
  return value.filter(isRecord)
}
